COMP-1175: Update HTTP interface specification to match the new design

thaapasa opened the pull request:
Update architecture model based on the new design. This will break backwards-compatibility with older versions.

kukushechkin commented on tests/conftest.py:
```diff
@@ -393,7 +393,7 @@ def sample_quantum_architecture():
             'qubits': ['QB1', 'QB2'],
             'qubit_connectivity': [['QB1', 'QB2']],
-            'operations': ['phased_rx', 'CZ'],
+            'operations': {'phased_rx': [['QB1'], ['QB2']], 'CZ': [['QB1', 'QB2']]},
```
There should also be ``Circuit`` instruction supported
qubits validator and tests verifying if we actually check them, similar to arty.

thaapasa replied:
Based on our discussion in Mattermost I've now changed the submit operation to fetch (possibly cached) quantum architecture and perform validation for the targeted qubits.

kukushechkin commented on src/iqm/iqm_client/iqm_client.py:
```diff
@@ -174,6 +174,10 @@
         'arity': 2,
         'args': {},
     },
+    'move': {
```
Supported instructions descriptions are listed in the docstrings at the top of the file, `move` should also be there.

thaapasa replied:
Agreed! I'm contacting William and asking him for help with documenting the gate correctly.

kukushechkin commented:
I would also add a changeling with the major version change that will probably be used once this is merged into the mainstream.

smite commented on src/iqm/iqm_client/iqm_client.py:
```diff
@@ -528,8 +532,8 @@ class QuantumArchitectureSpecification(BaseModel):
     """name of the quantum architecture"""
-    operations: list[str] = Field(...)
-    """list of operations supported by this quantum architecture"""
+    operations: dict[str, list[list[str]]] = Field(...)
+    """the operations supported by this quantum architecture, mapped to the allowed connections"""
```
```suggestion
    operations: dict[str, list[tuple[str, ...]]] = Field(...)
    """operations supported by this quantum architecture, mapped to the allowed loci"""
```

thaapasa replied:
Thanks! Fixed the comment

thaapasa commented:
> I would also add a changeling with the major version change that will probably be used once this is merged into the mainstream.

I've added some initial log entry that describes the changes generally

thaapasa commented:
The main iqm_client file has become quite large, so I extracted quantum_architecture.py and instruction.py from it and added some supporting operations there. If you wish to keep all these together in a single file (as it was), let me know, I can merge them back in. It looked to me that we could benefit from splitting the code into smaller pieces.

smite commented on src/iqm/iqm_client/quantum_architecture.py:
```diff
@@ -0,0 +1,103 @@
+classes.
+
+For more information on the topic, refer to the more comprehensive documentation in iqm_client.py.
+"""
```
```suggestion
"""
from __future__ import annotations
```

thaapasa replied:
Ah, thank you, fixed!

smite commented on src/iqm/iqm_client/quantum_architecture.py:
```diff
@@ -0,0 +1,103 @@
+    qubit_connectivity: list[list[str]] = Field(...)
+    """Qubit connectivity of this quantum architecture."""
+
+    def has_equivalent_operations(self, other: 'QuantumArchitectureSpecification'):
```
```suggestion
    def has_equivalent_operations(self, other: QuantumArchitectureSpecification):
```

smite commented on src/iqm/iqm_client/util.py:
```diff
@@ -43,4 +45,17 @@ def to_json_dict(obj: dict[str, Any]) -> dict:
+
+    Returns:
+        a copy of the list, in sorted order."""
+    r = list(input_list)
+    r.sort()
+    return r
```
```suggestion
    return input_list.copy().sort()
```

thaapasa replied:
Can't 😞 
The whole point for this utility function was that the sort() method doesn't return anything so I need to call it separately.

thaapasa replied:
Ah, actually, there's the `sorted` builtin I can use so I'll just ditch this.

smite commented on src/iqm/iqm_client/iqm_client.py:
```diff
@@ -98,6 +99,35 @@
+MOVE
+----
+
+The MOVE operation is an operation between a qubit and resonator than moves the
+1 state between the qubit and resonator. Specifically, it transforms 01 to 10
+and 10 to 01, where indexes are ordered qubit then resonator. However, the MOVE
+operation is not defined when both the resonator and qubit are in the 1 state;
+it is not defined on the 11 state. If the MOVE operation is applied to the
+11 state there will be substantial (roughly 63%) leakage into the 02 state.
+
+With 11 input undefined, the matrix form of the MOVE operation is:
+
+.. math:: \text{MOVE} = \begin{bmatrix}
+    1 & 0 & 0 & n/a \\
+    0 & 0 & 1 & n/a \\
+    0 & 1 & 0 & n/a \\
+    0 & 0 & 0 & n/a
+    \end{bmatrix}
+
+To ensure that both the qubit and resonator are not both in the 1 state, it is
+recommended that no single qubit gates are applied to the qubit in between a
+pair of MOVE operations.
```
```suggestion
The MOVE operation is a unitary population exchange operation between a qubit and a resonator.
Its effect is only defined in the invariant subspace :math:`S = \text{span}\{|00\rangle, |01\rangle, |10\rangle\}`,
where it swaps the populations of the states :math:`|01\rangle` and :math:`|10\rangle`.
Its effect on the orthogonal subspace is undefined.

MOVE has the following presentation in the subspace :math:`S`:

.. math:: \text{MOVE}_S = |00\rangle \langle 00| + a |10\rangle \langle 01| + a^{-1} |01\rangle \langle 10|,

where :math:`a` is an undefined complex phase that is canceled when the MOVE gate is applied a second time.

To ensure that the state of the qubit and resonator has no overlap with :math:`|11\rangle`, it is
recommended that no single qubit gates are applied to the qubit in between a
pair of MOVE operations.
```

thaapasa replied:
Looks good to me, but I don't know about the domain so I forwarded this to William to ask if he has any comments, but he might not be working today. I'll get back on this next week.

smite commented on src/iqm/iqm_client/iqm_client.py:
```diff
@@ -98,6 +99,35 @@
+.. code-block:: python
+   :caption: Example
+
+   Instruction(name='move', qubits=('alice', 'bob'), args={})
```
```suggestion
   Instruction(name='move', qubits=('alice', 'resonator'), args={})
```

jkotilahti commented on src/iqm/iqm_client/iqm_client.py:
```diff
@@ -888,6 +756,103 @@ def submit_circuits(
+            for instr in circuit.instructions:
+                IQMClient._validate_instruction(architecture, instr, qubit_mapping)
+
+    @staticmethod
+    def _validate_instruction(
+        architecture: QuantumArchitectureSpecification,
+        instruction: Instruction,
+        qubit_mapping: Optional[dict[str, str]] = None,
+    ):
```
The way `_validate_instruction` currently works does not easily allow `measure` instructions that operate on multiple qubits, since all of those multi-qubit loci would have to be specified separately in the quantum architecture (not scalable for large number of qubits). For example, if we have `"measure": [["QB1"], ["QB2"], ["QB3"]]` in the quantum architecture, it is only possible to use `measure` instructions with those single qubits but not e.g. `qubits=("QB1", "QB2")` (it would require `"measure": [["QB1"], ["QB2"], ["QB3"], ["QB1", "QB2"]]` in the architecture).

I think the validation should be modified so that for `measure` instructions, any locus which only contains qubits that are included in the single-qubit loci for `measure` in the architecture is allowed. Then specifying just the single-qubit loci (e.g. `[["QB1"], ["QB2"], ["QB3"]]`) for `measure` in the quantum architecture would allow also any combination of those qubits (e.g. `qubits=("QB1", "QB2")`) as locus for a `measure` instruction. Documentation and tests should also be updated accordingly.

Even if further changes to the architecture format are made, I think we will want the `measure` operations to work in essentially this way (list of single qubits in architecture, any combinration allowed for instruction locus), so I think changing it already would make sense.

thaapasa replied:
Sounds good, changed as suggested. All allowed qubits are now valid in the locus of the measure operation

jkotilahti commented on tests/test_validation.py:
```diff
@@ -0,0 +1,103 @@
+
+
+@pytest.mark.parametrize('qubits', [['QB1', 'COMP_R'], ['COMP_R'], ['QB1', 'QB2', 'QB4'], ['QB4']])
+def test_disallowed_measure_qubits(sample_move_architecture, qubits):
+    """
+    Tests that instruction validation succeeds for loci that are any combination of valid measure qubits
+    """
```
```suggestion
def test_disallowed_measure_qubits(sample_move_architecture, qubits):
    """
    Tests that instruction validation fails for loci containing any qubits that are not valid measure qubits
    """
```

thaapasa replied:
Thanks, missed this copy-pasta 🤦
